hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-router/src/test/java/org/apache/hadoop/yarn/server/router/secure/AbstractSecureRouterTest.java

```diff
@@ -0,0 +1,265 @@
+      if (mockRMs.containsKey(sc)) {
+        continue;
+      }
+      MockRM mockRM = new TestRMRestart.TestSecurityMockRM(conf);
+      mockRM.start();
+      mockRM.registerNode("127.0.0.1:1234", 8 * 1024, 4);
+      mockRMs.put(sc, mockRM);
+    }
+  }
+
+  /**
+   * Create the keytab for the given principal, includes
+   * raw principal and $principal/localhost.
+   *
+   * @param principal principal short name.
+   * @param filename filename of keytab.
+   * @return file of keytab.
+   * @throws Exception an error occurred.
+   */
+  public static File createKeytab(String principal, String filename) throws Exception {
+    Assert.assertTrue("empty principal", StringUtils.isNotBlank(principal));
+    Assert.assertTrue("empty host", StringUtils.isNotBlank(filename));
+    Assert.assertNotNull("Null KDC", kdc);
+    File keytab = new File(kdcWorkDir, filename);
+    kdc.createPrincipal(keytab,
+        principal,
+        principal + "/localhost",
+        principal + "/127.0.0.1");
+    return keytab;
+  }
+
+  /**
+   * Start the router in safe mode.
+   *
+   * @throws Exception an error occurred.
+   */
+  public synchronized void startSecureRouter(Boolean initRM) throws Exception {
+    Assert.assertNull("Router is already running", router);
+    UserGroupInformation.setConfiguration(conf);
+    router = new Router();
+    router.init(conf);
+    router.start();
+
+    if (initRM) {
+
+      setupSecureMockRM();
+
+      RouterClientRMService rmService = router.getClientRMProxyService();
+      RouterClientRMService.RequestInterceptorChainWrapper wrapper = rmService.getInterceptorChain();
+      FederationClientInterceptor interceptor = (FederationClientInterceptor) wrapper.getRootInterceptor();
+      FederationStateStoreFacade stateStoreFacade = interceptor.getFederationFacade();
+      FederationStateStore stateStore = stateStoreFacade.getStateStore();
+      FederationStateStoreTestUtil stateStoreUtil = new FederationStateStoreTestUtil(stateStore);
+      subClusters = new ArrayList<>();
+
+      for (int i = 0; i < NUM_SUBCLUSTER; i++) {
+        SubClusterId sc = SubClusterId.newInstance(Integer.toString(i));
+        stateStoreUtil.registerSubCluster(sc);
+        subClusters.add(sc);
+      }
+
+      Map<SubClusterId, ApplicationClientProtocol> clientRMProxies =
+          interceptor.getClientRMProxies();
+      for (Map.Entry<SubClusterId, MockRM> entry : mockRMs.entrySet()) {
+        SubClusterId keySubClusterId = entry.getKey();
+        if (clientRMProxies.containsKey(keySubClusterId)) {
+          continue;
+        }
+        MockRM mockRM = entry.getValue();
+        clientRMProxies.put(keySubClusterId, mockRM.getClientRMService());
+      }
+
+      MockRM firstRM = mockRMs.entrySet().stream().findFirst().get().getValue();
+      RouterRMAdminService routerRMAdminService = router.getRmAdminProxyService();
+      RouterRMAdminService.RequestInterceptorChainWrapper rmAdminChainWrapper =
```

Comment:
We had to expose a lot of these methods for the tests.
Is there a better way? Some mock or test class extension?

Reply:
> Some mock or test class extension?
This is a good idea, I will refactor this part of the code.